contributors/design-proposals/defer-containers.md

```diff
@@ -0,0 +1,248 @@
+rolled out as alpha as annotations based on its usefulness and community feedback it will graduate to PodSpec.
+
+## Examples 
+### Cleanup
+```yaml
+pod:
+  spec:
+    terminationGracePeriod: 60
+    initContainers: ...
+    containers: ...
+    deferContainers:
+#pre exit operations
+    - name: fsync
+      image: my-utils
+#Something like this https://docs.mongodb.com/manual/reference/command/fsync/
+      comamnd: ["/bin/sh", "-C", "./syncDisk.sh"] 
+      name: killApp
+      image: my-utils
+      command: ["/bin/sh", "-C", "./killAndWait.sh", "--name=${POD_NAME}"]  
+#post exit operation
+    - name: rm-tmpdir
+      image: my-utils
+      command: ["/bin/sh", "-C", "./disk_cleanup.sh"]
+#contact and inform a thirdparty system about this pods termination, such as reducing a refernce counter (if one is maintained)
+      name: ref-counter
+      image: my-utils
+      comamnd: ["/bin/sh", "-C", "./decrementRefCount.sh"]
+      ...
+```
+
+### Master-slave / Leader-follower statefulset down-size / scale down the replicas
+Below scripts 'selectaMaster.sh and reConfSlaves.sh should be designed in such a way that even if a terminating pod is a slave it
+shoudnt affect the cluster. This will fit controllers such as Statefulset because they gaurentee only one pod goes down at once.
+```yaml
+pod:
+  spec:
+    terminationGracePeriod: 60
+    initContainers: ...
+    containers: ...
+    deferContainers:
+    - name: electMaster
+      image: my-utils
+      #Select a mastter among the available slaves
+      comamnd: ["/bin/sh", "-C", "./selectAMaster.sh"] 
+      name: reconfigureSlaves
+      image: my-utils
+      #Re-configure all the slaves to the new master
+      command: ["/bin/sh", "-C", "./reconfSlaves.sh", "--MasterSlave=${MasterEP}"]  
+      name: killApp
+      image: my-utils
+      command: ["/bin/sh", "-C", "./killAndWait.sh", "--name=${POD_NAME}"] 
+      ...
+```
+### Shutdown Sequence
+If we attempt to shutdown a runing Oracle instance which has four stages https://docs.oracle.com/cd/B28359_01/backup.111/b28273/rcmsynta045.htm#RCMRF155
+```yaml
+pod:
+  spec:
+    initContainers: ...
+    containers: ...
+    deferContainers:
+#Pre-Exit
+    - name: shutdown-db
+      image: db-utils
+      #Select the shutdown sequnce 
+      comamnd: ["/bin/sh", "-C", "./shutDown.sh", "--shutdowntype=${SH_TYPE}"] 
+      name: waitforDB
+      image: db-utils
+      #run a script that will wait until the DB is down.
+      command: ["/bin/sh", "-C", "./waitForDB.sh"]  
+      ...
+```
+
+## Caviate
+This Design primarily focuses on handling graceful termination cases.  If the Node running a deferContainer configured pod
+crashes abruptly, then this design does not guarantee that cleanup was performed gracefully.  This still requires community
+feedback on how such scenarios are handled and how important it is for deferContainers to handle that situation. 
```

Comment:
I think adding this without any idea how to solve that problem is bad idea, it may create some illusion that some functionality is provided, while on production it may work not as "expected". ( compared with initContainers).